[gollum] New Chart

chris-sanders opened the pull request:
Signed-off-by: Chris Sanders <sanders.chris@gmail.com>

<!--
Before you open the request please review the following guidelines and tips to help it be more easily integrated:

- Describe the scope of your change - i.e. what the change does.
- Describe any known limitations with your change.
- Please run any tests or examples that can exercise your modified code.

Thank you for contributing! We will try to test and integrate the change as soon as we can. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

Also don't be worried if the request is closed or not integrated sometimes our priorities might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
-->

**Description of the change**

Adds a chart for hosting Gollum Wiki

**Benefits**

Wikijs has an option to use a git repository, and I tried to get that to fit my use case. However, the markdown files are only part of the data, a database is used for titles and tags. Also the markdown is a very strict interpretation that wouldn't work with my existing notes.

Gollum Wiki is used to power Github and Gitlab, it's markdown syntax supports many flavors, and it supports front matter for page specific configurations. Everything managed from git and markdown. I'm using it with a Vimwiki repository and I believe it supports Org-mode as well. If you use markdown files for note taking there's a good chance Gollum can host it for you.

**Possible drawbacks**

Access control not yet implemented, although Ingress can handle this and might be a better choice than using a Gollum specific solution. Gollum does support a list of plugins for authentication if that is added in the future.

**Applicable issues**

N/A

**Additional information**

N/A

**Checklist** <!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [X] Title of the PR starts with chart name (e.g. `[home-assistant]`)
- [X] Variables are documented in the README.md (this can be done with using our helm-docs wrapper `./hack/gen-helm-docs.sh stable <chart>`)

<!-- Keep in mind that if you are submitting a new chart, try to use our [common](https://github.com/k8s-at-home/charts/tree/master/charts/common) library as a dependency. This will help maintaining charts here and keep them consistent between each other -->

dirtycajunrice1 commented on charts/stable/gollum/templates/_container.tpl:
```diff
@@ -0,0 +1,78 @@
+  {{ toYaml . | nindent 2 }}
+  {{- end }}
+  {{- end }}
+  - {{ "--ref" }}
+  - {{ .Values.init.gitBranch }}
```
Redefining the entirety of common.controller.mainContainer for the inclusion of one item is not the way to go. This is designed to immediately accrue drift.
As you can see from line 8, values.command is already passable and this can be done in a common.yaml by modifying the end resulted .Values

dirtycajunrice1 commented on charts/stable/gollum/templates/_pod.yaml:
```diff
@@ -0,0 +1,78 @@
+  {{- toYaml . | nindent 2 }}
+{{- end }}
+enableServiceLinks: {{ .Values.enableServiceLinks }}
+initContainers:
+{{- with .Values.initContainers }}
+  {{- toYaml . | nindent 2 }}
+{{- end }}
+  - name: git-init
+    image: alpine/git
+    args:
+      - clone 
+      - -b
+      - {{ .Values.init.gitBranch  }}
+      - {{ .Values.init.gitUrl  }}
+      - /wiki
+    volumeMounts:
+      - mountPath: /wiki
+        name: wiki
```
All init containers can also be defined without redefining the common.controller.pod template. This can be done in a common.yaml as well and injected.

dirtycajunrice1 commented on charts/stable/gollum/templates/configmap-git-sync.yaml:
```diff
@@ -0,0 +1,17 @@
+  namespace: {{ .Release.Namespaces }} 
+data: 
+  git-sync: |
+{{ .Values.gitSync | indent 4 }}
```
```suggestion
    {{- .Values.gitSync | nindent 4 }}
```

dirtycajunrice1 commented on charts/stable/gollum/templates/configmap-git-sync.yaml:
```diff
@@ -0,0 +1,17 @@
+kind: ConfigMap
+metadata:
+  name: git-sync
+  namespace: {{ .Release.Namespaces }} 
```
Defining release namespace enforces something that is already set via helm. Not defining it allows postprocessors to make changes as needed.
```suggestion
```

dirtycajunrice1 commented on charts/stable/gollum/templates/configmap-git-sync.yaml:
```diff
@@ -0,0 +1,17 @@
+kind: ConfigMap
+metadata:
+  name: gollum-config
+  namespace: {{ .Release.Namespaces }} 
```
Defining release namespace enforces something that is already set via helm. Not defining it allows postprocessors to make changes as needed.
```suggestion
```

dirtycajunrice1 commented on charts/stable/gollum/templates/configmap-git-sync.yaml:
```diff
@@ -0,0 +1,17 @@
+  namespace: {{ .Release.Namespaces }} 
+data: 
+  config.rb: |
+{{ .Values.gollumConfig | indent 4 }}
```
```suggestion
    {{- .Values.gollumConfig | nindent 4 }}
```

dirtycajunrice1 commented on charts/stable/gollum/values.yaml:
```diff
@@ -0,0 +1,106 @@
+# -- For daemonset see persistence values, RWX will be required
+controllerType: statefulset
+# -- Start command
+command: ["gollum", "/wiki", "--config", "/config/config.rb", "--h1-title"]
```
This should be moved outside of the values.yaml to ensure it does not infer it should be changed.

dirtycajunrice1 commented on charts/stable/gollum/values.yaml:
```diff
@@ -0,0 +1,106 @@
+    # -- Enable PVC for the wiki repository. If no persistent storage is needed
+    # configure this as an emptyDir mount, it is used to share paths between
+    # containers.
+    enabled: true
```
persistence is opt in not out
```suggestion
    enabled: false
```

dirtycajunrice1 commented on charts/stable/gollum/values.yaml:
```diff
@@ -0,0 +1,106 @@
+    size: 2Gi
+
+# -- Mount gollum config into container
+additionalVolumeMounts:
+  - mountPath: /config
+    name: gollum-config
+
+additionalVolumes:
+  # -- Git sync command via configmap
+  - name: git-sync
+    configMap:
+      name: git-sync
+      defaultMode: 0755
+      items:
+        - key: git-sync
+          path: git-sync
+  # -- Gollum config via configmap
+  - name: gollum-config
+    configMap:
+      name: gollum-config
+      defaultMode: 0755
+      items:
+        - key: config.rb
+          path: config.rb
```
This should all be done outside of the values.yaml and injected as they seem needed. If they are optional it does not infer that from the configmap definitions.

dirtycajunrice1 commented on charts/stable/gollum/values.yaml:
```diff
@@ -0,0 +1,106 @@
+         committer.wiki.repo.git.push('origin', committer.wiki.ref)
+    end
+
+additionalContainers:
+  # -- Crond for periodic updates
+  - name: cron
+    image: alpine/git
+    command: ["crond", "-f", "-l", "8"]
+    volumeMounts:
+      - mountPath: /etc/periodic/15min/
+        name: git-sync
+      - mountPath: /wiki
+        name: wiki
+
+initContainers:
+# -- Clear on startup for fresh sync
+- name: clear-wiki
+  image: alpine
+  command:
+    - sh
+    - -c
+    - rm -fr /wiki/*; rm -fr /wiki/.*; ls -la /wiki/;
+  volumeMounts:
+    - mountPath: /wiki
+      name: wiki
```
These should be moved outside of values.yaml  and injected if not optional.

dirtycajunrice1 commented on charts/stable/gollum/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
+appVersion: 5.X
```
app version should match docker tagged version.

dirtycajunrice1 reviewed (changes requested):
Welcome to the k8s@home community and thanks for your first PR! There are some design changes needed and i have noted them throughout. Additionally, You can use a single configmap and achieve the same result by key reference without needing to make 2 separate.

onedr0p commented on charts/stable/gollum/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
+appVersion: 5.X
+description: gollum helm package
```
```suggestion
description: Gollum is a simple wiki system built on top of Git
```

chris-sanders commented:
Looking at the history, it looks like you're using a rebase workflow so I've rebased the changes. I believe I've addressed everything, it's a pretty substantial set of changes. Thanks for the guidance on using the common library, it's really nice now that I know how to use it.

chris-sanders commented on charts/stable/gollum/templates/configmap.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: git-sync
+data: 
+  git-sync: |
+    #!/bin/bash
+    cd /wiki && {{ .Values.gollum.syncCommand}}
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: gollum-config
+data: 
+  config.rb: |
+  {{- .Values.gollum.config | nindent 4 }}
```
Nope no reason, I just missed this among the other changes. I'll merge this and go change the references to consolidate.

chris-sanders commented:
It looks like I need to add someone specific to the maintainers for lint. If you can clue me in on what that needs to be.
I'm not seeing any info on the install test at all. Is there any way for me to see more info on that, or reproduce locally so I can address it?

dirtycajunrice reviewed (dismissed):
Minor cleanup can still happen, but cleaned up enough for 1.0

bjw-s commented on charts/stable/gollum/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+- https://github.com/gollum/gollum
+- https://github.com/gollum/docker
+maintainers:
+- name: Chris Sanders
```
```suggestion
- name: @chris-sanders
```
`ct` apparently wants to have an actual GitHub id here
